tests/src/test/java/org/eclipse/hono/tests/amqp/EventAmqpIT.java

```diff
@@ -56,4 +65,65 @@ protected void assertAdditionalMessageProperties(final VertxTestContext ctx, fin
             assertThat(msg.isDurable()).isTrue();
         });
     }
+
+    /**
+     * Verifies that an event message from a device has been successfully sent and a north bound application, 
+     * which connects after the event has been sent, can successfully receive those event message.
+     *
+     * @param ctx The vert.x test context.
+     * @throws InterruptedException if test execution gets interrupted.
+     */
+    @Test
+    public void testEventMessageAlreadySentIsDeliveredWhenConsumerConnects(final VertxTestContext ctx)
+            throws InterruptedException {
+        final VertxTestContext setup = new VertxTestContext();
+        final String tenantId = helper.getRandomTenantId();
+        final String deviceId = helper.getRandomDeviceId(tenantId);
+
+        setupProtocolAdapter(tenantId, deviceId, ProtonQoS.AT_LEAST_ONCE, false)
+                .map(s -> sender = s)
+                .onComplete(setup.completing());
+
+        assertThat(setup.awaitCompletion(5, TimeUnit.SECONDS)).isTrue();
+        if (setup.failed()) {
+            ctx.failNow(setup.causeOfFailure());
+            return;
+        }
+
+        // WHEN a device that belongs to the tenant publishes an event
+        final AtomicInteger receivedMessageCount = new AtomicInteger(0);
+        final Message event = ProtonHelper.message();
+        MessageHelper.setPayload(event, "opaque/binary", Buffer.buffer("hello"));
+        event.setAddress(getEndpointName());
+        final Promise<?> sendingComplete = Promise.promise();
+        sender.send(event, delivery -> {
+            if (Accepted.class.isInstance(delivery.getRemoteState())) {
+                sendingComplete.complete();
+            } else {
+                sendingComplete.fail(AmqpErrorException.from(delivery.getRemoteState()));
+            }
+        });
+
+        sendingComplete
+                .future()
+                .compose(eventSent -> {
+                    //THEN create a consumer once the event message has been successfully sent
+                    final Promise<MessageConsumer> consumerCreated = Promise.promise();
+                    vertx.setTimer(4000, tid -> {
+                        log.info("opening event consumer for tenant [{}]", tenantId);
+                        createConsumer(tenantId, msg -> receivedMessageCount.incrementAndGet())
+                                .onComplete(consumerCreated);
+                    });
+                    return consumerCreated.future();
+                })
+                .compose(consumer -> {
+                    final Promise<Void> done = Promise.promise();
+                    vertx.setTimer(1000, tid -> {
+                        //THEN verify if the message is received by the consumer
+                        assertThat(receivedMessageCount.get()).isEqualTo(1);
```

Comment:
instead of verifying the number of messages here I'd rather assert a particular property of the message that has been sent in the consumer itself, before flagging the checkpoint ...

Reply:
I have made use of a random UUID as a payload message. The event consumer verifies the payload content with that of the random UUID value.